src/zelos/tools/zdbserver/test_zdbserver.py

```diff
@@ -0,0 +1,206 @@
+        )
+
+        return srv, rpc
+
+    def stop_server(
+        self, srv: multiprocessing.Process, rpc: xmlrpc.client.ServerProxy
+    ) -> None:
+        rpc.server_shutdown()
+        srv.join(5)
+        self.assertFalse(srv.is_alive())
+        if srv.is_alive():
+            srv.terminate()
+
+    # @pytest.mark.usefixtures("serial")
+    # def test_server_connect(self):
+    #     proc, rpc = self.start_server()
+    #     self.stop_server(proc, rpc)
+
+    @pytest.mark.usefixtures("serial")
+    def test_server_api(self):
+        # Do this all in one test so we're not creating many
+        # sub-processes.
+        proc, zdb = self.start_server()
+
+        zdb_exception = None
+
+        try:
+            self.assertEqual(
+                zdb.get_filepath(),
+                path.join(DATA_DIR, "static_elf_helloworld"),
+            )
+
+            # Test syscall breaks
+            zdb.set_syscall_breakpoint("brk")
+            break_state = zdb.run()
+            del break_state["syscall"]["retval"]  # address may change
+            self.assertEqual(
+                str(break_state),
+                "{'pc': '0x815b577', 'syscall': {'name': 'brk', 'args': "
+                "[{'type': 'void*', 'name': 'addr', 'value': '0x0'}], "
+                "'retval_register': 'eax'}, 'bits': 32}",
+            )
+
+            # Test syscall break argument values
+            break_state = zdb.run()
+            brk_target_addr = int(
+                break_state["syscall"]["args"][0]["value"], 16
+            )
+            # Address may change, so just check it is non-zero
+            self.assertNotEqual(0, brk_target_addr)
+
+            zdb.remove_syscall_breakpoint("brk")
+
+            # Test address breaks
+            zdb.set_breakpoint("0x080eccf7", False)
+            break_state = zdb.run()
+            self.assertEqual(
+                str(break_state),
+                "{'pc': '0x80eccf7', 'syscall': {}, 'bits': 32}",
+            )
+            zdb.remove_breakpoint("0x080eccf7")
+
+            # Test read/write register
+            pc = int(zdb.read_register("pc"), 16)
+            self.assertEqual(pc, 0x80ECCF7)
+            zdb.write_register("pc", "0xdeadbeef")
+            pc = int(zdb.read_register("eip"), 16)
+            self.assertEqual(pc, 0xDEADBEEF)
+            zdb.write_register("eip", "0x80eccf7")
+            pc = int(zdb.read_register("pc"), 16)
+            self.assertEqual(pc, 0x80ECCF7)
+
+            ecx = int(zdb.read_register("ecx"), 16)
+            self.assertEqual(ecx, 0x81E9CA0)
+
+            # Test `zdbserver` exceptions
+            # ...
+
+            # Test memory read after break
+            stack_var = int(zdb.read_register("esp"), 16) + 0x1C
+            data = zdb.read_memory(f"0x{stack_var:x}", 4)
+            print(data.data)
+            # self.assertEqual(data.data, b"\xc0\x00\x00\x90")  # after inst
+            self.assertEqual(data.data, b"\x80\x00\x00\x90")  # before inst
```

Comment:
Interesting. This test fails only if this test script resides in the zdbserver directory. Same test passes if in the tests/tools directory.

Reply:
It was the differing host path affecting emulated execution. Fixed by ensuring virtual path is used in Linux emulation.